### PR TITLE
[CodeGenPrepare] Check types when unmerging GEPs across indirect branches

### PR DESCRIPTION
**Pull request by momo5502**
The optimization in CodeGenPrepare, where GEPs are unmerged across indirect branches must respect the types of both GEPs and their sizes when adjusting the indices.

The sample here shows the bug:

https://godbolt.org/z/8e9o5sYPP

The value `%elementValuePtr` addresses the second field of the `%struct.Blub`. It is therefore a GEP with index 1 and type i8.
The value `%nextArrayElement` addresses the next array element. It is therefore a GEP with index 1 and type `%struct.Blub`.

Both values point to completely different addresses, even if the indices are the same, due to the types being different.
However, after CodeGenPrepare has run, `%nextArrayElement` is a bitcast from `%elementValuePtr`, meaning both were treated as equal.

The cause for this is that the unmerging optimization does not take types into consideration.
It sees both GEPs have `%currentArrayElement` as source operand and therefore tries to rewrite `%nextArrayElement` in terms of `%elementValuePtr`.
It changes the index to the difference of the two GEPs. As both indices are `1`, the difference is `0`.  As the indices are `0` the GEP is later replaced with a simple bitcast in CodeGenPrepare.

Before adjusting the indices, the types of the GEPs would have to be aligned and the indices scaled accordingly for the optimization to be correct.
Due to the size of the struct being `16` and the `%elementValuePtr` pointing to offset `1`, the correct index for the unmerged `%nextArrayElement` would be 15.

I assume this bug emerged from the opaque pointer change as GEPs like `%elementValuePtr` that access the struct field based of type i8 did not naturally occur before.

In light of future migration to ptradd, simply not performing the optimization if the types mismatch should be sufficient.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
High level comment: Would bailing out of the transform if the source element types don't match be "good enough"?

I'd prefer not to add the extra logic if we don't particularly care about making it work with mismatching GEP types. (Longer term this will be automatically fixed by migration to ptradd.)

File: llvm/test/CodeGen/Generic/indirect-br-gep-unmerge.ll [hunk 1]
```diff
@@ -0,0 +1,29 @@
+; RUN: llc %s -stop-after=codegenprepare -o - | FileCheck %s
```
[the remark this answers is not in the document]
Comment (momo5502) [reply llvm/test/CodeGen/Generic/indirect-br-gep-unmerge.ll]: done

**Review by nikic — COMMENTED** [review]
This looks good, just some test nits.

File: llvm/test/CodeGen/Generic/indirect-br-gep-unmerge.ll [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+; CHECK-NEXT:    [[MERGE:%.*]] = phi ptr [ [[ELEMENTVALUEPTR]], [[ENDBLOCK:%.*]] ], [ [[RETVAL_CLONE:%.*]], [[ENDBLOCK_CLONE]] ]
+; CHECK-NEXT:    ret ptr [[MERGE]]
+; CHECK:       endBlock.clone:
+; CHECK-NEXT:    [[RETVAL_CLONE]] = phi ptr [ [[ARRAY]], [[ENTRY]] ], [ [[ELEMENTVALUEPTR]], [[LOOPFOOTER]] ]
+; CHECK-NEXT:    br label [[DOTSPLIT]]
+;
+entry:
+  br i1 %skip, label %loopHeader, label %endBlock
+
+loopHeader:                                                ; preds = %2, %1
```
[the remark this answers is not in the document]
Comment (momo5502) [reply llvm/test/CodeGen/Generic/indirect-br-gep-unmerge.ll]: done

File: llvm/test/CodeGen/Generic/indirect-br-gep-unmerge.ll [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+; RUN: opt -S -codegenprepare %s -o - | FileCheck %s
+
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-unknown-linux-gnu"
+
+%struct.Blub = type { i8, i8, ptr }
+
+@indirectBrPtr = external hidden global ptr
+
+define dso_local noundef ptr @testFunc(ptr noundef readonly %array, i1 %skip) {
```
[the remark this answers is not in the document]
Comment (momo5502) [reply llvm/test/CodeGen/Generic/indirect-br-gep-unmerge.ll]: done

File: llvm/test/CodeGen/Generic/indirect-br-gep-unmerge.ll [hunk 4]
```diff
@@ -0,0 +1,51 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
+; RUN: opt -S -codegenprepare %s -o - | FileCheck %s
+
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-unknown-linux-gnu"
```
[the remark this answers is not in the document]
Comment (momo5502) [reply llvm/test/CodeGen/Generic/indirect-br-gep-unmerge.ll]: i don't think this is x86 specific, but i moved it to the x86 directory as it did not work without the triple.

**Comment by momo5502** [comment]
Thank you :)

**Pull request merged**